src/main/scala/firrtl/passes/ModuleNamePrefix.scala

```diff
@@ -0,0 +1,57 @@
+package firrtl.passes
+
+import firrtl._
+import firrtl.options.{Dependency, RegisteredTransform, ShellOption}
+import firrtl.transforms.DedupModules
+import firrtl.annotations.NoTargetAnnotation
+import firrtl.stage.RunFirrtlTransformAnnotation
+
+/** Specifies a global prefix for all module names. */
+case class ModuleNamePrefixAnnotation(prefix: String = "") extends NoTargetAnnotation
+
+class ModuleNamePrefixTransform extends Transform with DependencyAPIMigration with RegisteredTransform {
+
+  override def prerequisites = Seq(Dependency[DedupModules])
+  override def optionalPrerequisiteOf = Seq(Dependency[VerilogEmitter])
+
+  override def invalidates(a: Transform) = false
+
+  val options = Seq(
+    new ShellOption[String](
+      longOption = "module-name-prefix",
+      toAnnotationSeq = (a: String) =>
+        Seq(ModuleNamePrefixAnnotation(a), RunFirrtlTransformAnnotation(new ModuleNamePrefixTransform)),
+      helpText = "Add global prefix to every verilog module (default: \"\")",
+      shortOption = Some("prefix"),
+      helpValueName = Some("<prefix>")
+    )
+  )
+
+  private def onStmt(s: ir.Statement, prefix: String): ir.Statement = s match {
+    case i : ir.DefInstance => i.copy(module = prefix + i.module)
+    case other => other.mapStmt(onStmt(_, prefix))
+  }
+
+  private def onModule(m: ir.DefModule, prefix: String): ir.DefModule = m match {
+    case e : ir.ExtModule => e.copy(name = prefix + e.name)
+    case mod: ir.Module =>
+      val name = prefix + mod.name
+      val body = onStmt(mod.body, prefix)
+      mod.copy(name=name, body=body)
+  }
+
+  override protected def execute(state: CircuitState): CircuitState = {
+    val annos = state.annotations.collect{ case a: ModuleNamePrefixAnnotation => a.prefix }.distinct
+    annos match {
+      case Seq() =>
+        logger.info("[ModuleNamePrefixAnnotation] No ModuleNamePrefixAnnotation annotation found.")
+        state
+      case Seq("") => state
+      case Seq(prefix) =>
+        val c = state.circuit.mapModule(onModule(_, prefix))
```

Comment:
Related to the comment on instance renaming... Because FIRRTL allows use-before-def of instances, you can do the correct instance renaming by:
1. Topologically sorting the modules
2. Reversing the sort (so its now leaves to roots)
3. Renaming modules and instances
4. Converting back to the original order

This enables you to rename modules in the circuit namespace and guarantee that you'll rename them before you ever see an instance of them.

There is a related pass for renaming everything in the circuit which shows how to do this dance: https://github.com/chipsalliance/firrtl/blob/master/src/main/scala/firrtl/transforms/ManipulateNames.scala#L449

Reply:
Since we do not get any name collisions by construction (see my previous comment), the renaming is a completely deterministic/pure function and you should be able to run this is any order. If we want more features than just prefixing every module, we are probably going to end up with something more like the ManipulateNames transform that you mentioned.

Reply:
Keeping this simple would make sense. Good analysis, @ekiwi.